<div style="
    background-color: #7BAFD4;
    border-radius: 18px;
    padding: 22px 26px;
    display: flex;
    align-items: center;
    gap: 24px;
    margin-bottom: 18px;
">
    <img src="592_avatar_lr.png" alt="GEOG 592"
        style="width:110px;height:110px;border-radius:22px;">
    <div>
        <div style="
            font-family: Impact, 'Arial Black', Arial, sans-serif;
            font-size: 42px;
            font-weight: 900;
            letter-spacing: 2px;
            line-height: 1;
            color: white;
        ">GEOG 592</div>
        <div style="
            font-family: Arial, Helvetica, sans-serif;
            font-size: 22px;
            font-weight: 700;
            color: #13294B;
            margin-top: 8px;
        ">GIS Programming</div>
        <div style="
            font-family: Arial, Helvetica, sans-serif;
            font-size: 16px;
            color: #13294B;
            margin-top: 6px;
        ">Homework 4: Pandas Intro</div>
    </div>
</div>

In [5]:
import pandas as pd

print(pd.__version__)

filename = "data/NC_Fire_Stations.csv"
stations = pd.read_csv(filename)

3.0.5


In [6]:
stations.head()

,OBJECTID,FD_ID,DEPT_NAME,STATION_NUMBER,STATION_ADDRESS,CITY,STATE,COUNTY,ZIP_CODE,LATITUDE,LONGITUDE,x,y
0,31333,1807,Hickory Fire Department,2,1305 9th Ave NE,Hickory,NC,Catawba,28601,35.741326,-81.312764,-81.312764,35.741326
1,31334,1807,Hickory Fire Department,3,135 11th Street NW,Hickory,NC,Catawba,28601,35.734400,-81.357300,-81.357300,35.734400
2,31335,1807,Hickory Fire Department,4,3101 9th Avenue,Hickory,NC,Catawba,28601,35.742518,-81.387943,-81.387943,35.742518
3,31336,1807,Hickory Fire Department,5,1425 21st Street Drive SE,Hickory,NC,Catawba,28601,35.703636,-81.279084,-81.279084,35.703636
4,31337,1807,Hickory Fire Department,6,3036 North Center Street,Hickory,NC,Catawba,28601,35.773600,-81.328300,-81.328300,35.773600


The head method prints the first 5 rows of the dataset. 

In [7]:
stations.shape

(2126, 13)

The shape attribute shows that there are a total of 2126 rows and 13 different columns. 

In [8]:
stations.columns

Index(['OBJECTID', 'FD_ID', 'DEPT_NAME', 'STATION_NUMBER', 'STATION_ADDRESS',
       'CITY', 'STATE', 'COUNTY', 'ZIP_CODE', 'LATITUDE', 'LONGITUDE', 'x',
       'y'],
      dtype='str')

The columns attribute show the names of all of the columns in the dataset. 

In [9]:
stations.dtypes

OBJECTID             int64
FD_ID                int64
DEPT_NAME              str
STATION_NUMBER       int64
STATION_ADDRESS        str
CITY                   str
STATE                  str
COUNTY                 str
ZIP_CODE             int64
LATITUDE           float64
LONGITUDE          float64
x                  float64
y                  float64
dtype: object

The dtypes attribute shows the data types of all columns. For instance, there are integers, strings, and float data types in the fire stations csv file. 

In [10]:
stations.info()

<class 'pandas.DataFrame'>
RangeIndex: 2126 entries, 0 to 2125
Data columns (total 13 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   OBJECTID         2126 non-null   int64  
 1   FD_ID            2126 non-null   int64  
 2   DEPT_NAME        2126 non-null   str    
 3   STATION_NUMBER   2126 non-null   int64  
 4   STATION_ADDRESS  2126 non-null   str    
 5   CITY             2126 non-null   str    
 6   STATE            2126 non-null   str    
 7   COUNTY           2126 non-null   str    
 8   ZIP_CODE         2126 non-null   int64  
 9   LATITUDE         2125 non-null   float64
 10  LONGITUDE        2125 non-null   float64
 11  x                2125 non-null   float64
 12  y                2125 non-null   float64
dtypes: float64(4), int64(4), str(5)
memory usage: 216.1 KB


The info method provides helpful information showing the number of non-null values in each column and their data types. For this example, it found that there is one row that has all information except latitude, longitude, x, and y coordinates that register as null. This highlights where the csv needs to be cleaned. 

In [ ]:
valid_coordinates = stations["x"].notna() & (stations["x"] != "")
valid_coordinates.head() 
print('length before mask', len(stations)) 
print('length of mask is', len(valid_coordinates)) 
clean_x = stations[valid_coordinates] 
print('length of our new df after using the mask:', len(clean_x)) 
print('since we had', len(valid_coordinates[valid_coordinates == True]), 'True values') 
clean_x.head()

length before mask 2126
length of mask is 2126
length of our new df after using the mask: 2125
since we had 2125 True values


,OBJECTID,FD_ID,DEPT_NAME,STATION_NUMBER,STATION_ADDRESS,CITY,STATE,COUNTY,ZIP_CODE,LATITUDE,LONGITUDE,x,y
0,31333,1807,Hickory Fire Department,2,1305 9th Ave NE,Hickory,NC,Catawba,28601,35.741326,-81.312764,-81.312764,35.741326
1,31334,1807,Hickory Fire Department,3,135 11th Street NW,Hickory,NC,Catawba,28601,35.734400,-81.357300,-81.357300,35.734400
2,31335,1807,Hickory Fire Department,4,3101 9th Avenue,Hickory,NC,Catawba,28601,35.742518,-81.387943,-81.387943,35.742518
3,31336,1807,Hickory Fire Department,5,1425 21st Street Drive SE,Hickory,NC,Catawba,28601,35.703636,-81.279084,-81.279084,35.703636
4,31337,1807,Hickory Fire Department,6,3036 North Center Street,Hickory,NC,Catawba,28601,35.773600,-81.328300,-81.328300,35.773600


I had to look up the method that removes na values in panda (.notna), so that the mask could remove the one row with missing coordinate data, but I did not use any AI on this homework. Therefore, the mask iterated through each x value in the stations dataset and checked if it was not na and not empty, returning only the those cells. It cleaned the dataset by removing the row with na values. 